### PR TITLE
[OneExplorer] Refactor with abstract class

### PR DESCRIPTION
**Pull request by dayo09**
This commit refactors OneExplorer by using abstract class.
It redefines Node as an abstract class
and extends into each NodeType.

ONE-vscode-DCO-1.0-Signed-off-by: dayo09 <dayoung.lee@samsung.com>

---

From #1039

### CODE REVIEW

File: src/OneExplorer/OneExplorer.ts [hunk 1]
```diff
@@ -113,6 +146,179 @@ class Node {
   }
 }
 
+class NodeFactory {
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:149]: (1) Used factory pattern for creating nodes

File: src/OneExplorer/OneExplorer.ts [hunk 2]
```diff
@@ -80,17 +108,22 @@ enum NodeType {
   product,
 }
 
-class Node {
-  type: NodeType;
+abstract class Node {
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:111]: (2) Made Node as an abstract class (strategy pattern)

Each type of nodes are made into each classes

File: src/OneExplorer/OneExplorer.ts [hunk 3]
```diff
@@ -80,17 +108,22 @@ enum NodeType {
...
+  abstract readonly type: NodeType;
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:112]: To make sure that type is not altered

File: src/OneExplorer/OneExplorer.ts [hunk 4]
```diff
@@ -80,17 +108,22 @@ enum NodeType {
...
   childNodes: Node[];
   uri: vscode.Uri;
 
-  constructor(type: NodeType, childNodes: Node[], uri: vscode.Uri) {
-    this.type = type;
-    this.childNodes = childNodes;
+  abstract icon: vscode.ThemeIcon;
+  abstract openViewType: string|undefined;
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:117]: If openViewType is undefined, it means openView is not supported. Nothing will happen when user clicks the item (no view opened)

File: src/OneExplorer/ConfigObject.ts [hunk 5]
```diff
@@ -29,19 +29,56 @@ import {Logger} from '../Utils/Logger';
  */
 export interface Artifact {
   /**
-   * An artifact's type
+   * An artifact's attribute
    */
-  type: string;
+  attr: ArtifactAttr;
```
Comment (dayo09) [inline src/OneExplorer/ConfigObject.ts:34]: (3) Introduced ArtifactAttr interface

To deliver artifact details

File: src/OneExplorer/ConfigObject.ts [hunk 6]
```diff
@@ -29,19 +29,56 @@ import {Logger} from '../Utils/Logger';
...
+   *
+   * @reference vscode.openWith
+   */
+  openViewType?: string;
+
+  /**
+   * Hidden from the default view.
+   * The status can be unhide by command
+   */
+  canHide?: boolean;
```
Comment (dayo09) [inline src/OneExplorer/ConfigObject.ts:81]: The hide command will be introduced soon.

File: src/OneExplorer/ConfigObject.ts [hunk 7]
```diff
@@ -313,17 +350,17 @@ export class ConfigObj {
     let locatorRunner = new LocatorRunner();
 
     locatorRunner.register({
-      artifactAttr: {type: 'model', ext: '.tflite'},
+      artifactAttr: {ext: '.tflite'},
```
Comment (dayo09) [inline src/OneExplorer/ConfigObject.ts:353]: Replaced `type` with `artifactAttr`, which is set as `undefined` here

File: src/OneExplorer/OneExplorer.ts [hunk 8]
```diff
@@ -366,149 +572,29 @@ input_path=${modelName}.${extName}
...
-    this.searchConfig(baseModelNode);
-
-    return baseModelNode;
-  }
-
-  /**
-   * Get the list of .cfg files wiithin this workspace
-   * TODO Move to constructor
-   */
-  private getCfgList(root: string = this.workspaceRoot!.fsPath): string[] {
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:443]: Moved

File: src/OneExplorer/OneExplorer.ts [hunk 9]
```diff
@@ -22,9 +22,37 @@ import * as vscode from 'vscode';
...
 import {obtainWorkspaceRoot, RealPath} from '../Utils/Helpers';
 import {Logger} from '../Utils/Logger';
+import {ArtifactAttr, ConfigObj} from './ConfigObject';
 
-import {Artifact, ConfigObj} from './ConfigObject';
-import {OneccRunner} from './OneccRunner';
+/**
+ * Get the list of .cfg files wiithin the workspace
+ */
+function getCfgList(root: string = obtainWorkspaceRoot()): string[] {
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:30]: (4) Moved getCfgList helper function 

From [here](https://github.com/Samsung/ONE-vscode/pull/1041/files#r921786036)

File: src/OneExplorer/OneExplorer.ts [hunk 10]
```diff
@@ -113,6 +146,179 @@ class Node {
...
+      node = new BaseModelNode(uri);
+    } else if (type === NodeType.config) {
+      node = new ConfigNode(uri);
+    } else if (type === NodeType.product) {
+      node = new ProductNode(uri);
+    } else {
+      throw Error('Undefined NodeType');
+    }
+
+    node.buildChildren();
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:166]: Each nodes will make its child nodes based on its own strategy

File: src/OneExplorer/OneExplorer.ts [hunk 11]
```diff
@@ -124,16 +330,16 @@ export class OneNode extends vscode.TreeItem {
...
 
-    if (node.type === NodeType.config) {
-      this.iconPath = new vscode.ThemeIcon('gear');
-    } else if (node.type === NodeType.directory) {
-      this.iconPath = vscode.ThemeIcon.Folder;
-    } else if (node.type === NodeType.baseModel) {
-      this.iconPath = vscode.ThemeIcon.File;
-    } else if (node.type === NodeType.product) {
-      this.iconPath = vscode.ThemeIcon.File;
+    if (node.openViewType) {
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:333]: (5) Introduced openViewType attribute

To make each node's behavior clear

File: src/OneExplorer/OneExplorer.ts [hunk 12]
```diff
@@ -124,16 +330,16 @@ export class OneNode extends vscode.TreeItem {
...
+      this.command = {
+        command: 'vscode.openWith',
+        title: 'Open with Custom Viewer',
+        arguments: [node.uri, node.openViewType]
+      };
     }
 
+    this.iconPath = node.icon;
```
Comment (dayo09) [inline src/OneExplorer/OneExplorer.ts:341]: (5) Introduced icon attribute

To make each node's behavior clear

**Review by llFreetimell — DISMISSED** [review]

File: src/OneExplorer/OneExplorer.ts [hunk 13]
```diff
@@ -113,6 +146,179 @@ class Node {
...
+  }
+
+  /**
+   * Build a sub-tree under the node
+   *
+   * directory
+   *   ∟ baseModel
+   *      ∟ config      <- this
+   *         ∟ product  <- children
+   */
```
Comment (yunjayh) [inline src/OneExplorer/OneExplorer.ts:287]: Not to give a pressure to implement, but I'm wondering that do we have to make sure the product is generated by the parent node config or not. 
If two different configs generate two different product with same name, the product node will be presented in child of both configs, but one of them should be wrong. (wrong? fake? false?)
Was there any discussion about it before?

**Review by yunjayh — APPROVED** [review]
The PR itself looks good to me :-)

File: src/OneExplorer/OneExplorer.ts (hunk 13 above)
Comment (dayo09) [reply src/OneExplorer/OneExplorer.ts:287]: > If two different configs generate two different product with same name, the product node will be presented in child of both configs, but one of them should be wrong. (wrong? fake? false?)

There is no such case.

Each config file will have paths of its produced files.
If some files' path are the same, then it means the files are the same.
So you question is: If two different configs generates product with the same name, will the product node will presented as a child of both nodes?
And the answer is naturally, yes :-D

**Comment by dayo09** [comment]
@llFreetimell 

@yunjayh is on his vacation. It takes a week and I committed his suggestion. Could we proceed without his approval?

**Review by llFreetimell — APPROVED** [review]

**Review by YongseopKim — APPROVED** [review]
LGTM

**Comment by llFreetimell** [comment]
As @yunjayh is out of office this week, I will merge this PR without his approval.
(In addition, he actually approved this PR)

**Pull request merged**